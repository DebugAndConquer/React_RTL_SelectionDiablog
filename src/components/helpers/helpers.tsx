import { ISelectorWidgetOption } from "../../types/SelectorWidgetTypes";

// Immutably updates an isSelected attribute for the entry with provided value
export const setValuesWrapper = (
  value: number,
  isSelected: boolean,
  setter: (value: React.SetStateAction<ISelectorWidgetOption[]>) => void
): void => {
  setter((prev) => {
    return prev.map((el) => {
      // When the value matches the searched one -> update isSelected attribute
      if (el.value === value) {
        return {
          ...el,
          isSelected: isSelected,
        };
        // When no match -> leave as it is
      } else {
        return el;
      }
    });
  });
};

// Generates the specified number of list elements with name 'Element <x>'
export const generateSampleData = (
  numberOfEl: number
): ISelectorWidgetOption[] => {
  const sampleData: ISelectorWidgetOption[] = [];
  for (let i = 1; i <= numberOfEl; i++) {
    sampleData.push({ label: `Element ${i}`, value: i, isSelected: false });
  }
  return sampleData;
};

export const spacesToDash = (initialString: string): string =>
  initialString.replaceAll(" ", "-");
