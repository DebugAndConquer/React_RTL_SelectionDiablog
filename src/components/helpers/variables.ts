export const ELEMENT_COUNT = 300; // Total test elements to create

export const ACTIVE_ENTRY = "selector-dialog__entries__entry"; // Regular entry class
export const DISABLED_ENTRY_MOD = "selector-dialog__entries__entry--disabled"; // A modifier to make it disabled
export const MAXIMUM_SELECTED_ENTRIES = 3; // No more than specified number of entries will be allowed to be selected

// FILTERS
export const FILTER_VALUES = {
    FILTER_NONE: 0,
    FILTER_MORE_THAN_10: 1,
    FILTER_MORE_THAN_50: 2,
    FILTER_MORE_THAN_100: 3
}