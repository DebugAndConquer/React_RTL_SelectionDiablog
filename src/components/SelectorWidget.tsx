import React, { useEffect, useState } from 'react';
import { ISelectorWidgetOption, ISelectorWidgetProps } from "../types/SelectorWidgetTypes";
import { setValuesWrapper } from './helpers/helpers';
import SelectorDialog from './SelectorDialog';
import SelectorOptions from './SelectorOptions';

const SelectorWidget: React.FC<ISelectorWidgetProps> = ({ options }) => {
    const [values, setValues] = useState<ISelectorWidgetOption[]>([]);
    const [showDialog, setShowDialog] = useState<boolean>(false);

    useEffect(() => {
        setValues(options);
    }, [options])

    const saveData = (newValues: ISelectorWidgetOption[]): void => {
        setValues(newValues);
        setShowDialog(false);
    }

    return (
        <div className='sel-widget'>
            <h1>Element selection</h1>
            <p>Number of selected elements: {values.filter((opt) => opt.isSelected).length}</p>
            <SelectorOptions values={values} handleUnselect={(value) => setValuesWrapper(value, false, setValues)} />
            <br />
            <button className='btn btn--green' onClick={() => setShowDialog(!showDialog)}>Change my selection</button>
            {
                showDialog &&
                <SelectorDialog
                    options={values}
                    onSave={saveData}
                    onHide={() => setShowDialog(false)} />
            }
        </div >
    )
}

export default SelectorWidget;