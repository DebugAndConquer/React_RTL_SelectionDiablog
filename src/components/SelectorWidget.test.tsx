import React from "react";
import { render, screen } from "@testing-library/react";
import SelectorWidget from "./SelectorWidget";
import { generateSampleData } from "./helpers/helpers";
import userEvent from "@testing-library/user-event";

test("Verifies that the dialog opens, its content is visible and then matches the snapshot", async () => {
  const { container } = render(
    <SelectorWidget options={generateSampleData(10)} />
  );
  const modalButtonElement = screen.getByText(/Change my selection/i);
  await userEvent.click(modalButtonElement);
  const saveButtonElement = screen.getByText(/Save/i);
  expect(saveButtonElement).toBeInTheDocument();
  expect(container).toMatchSnapshot();
});
