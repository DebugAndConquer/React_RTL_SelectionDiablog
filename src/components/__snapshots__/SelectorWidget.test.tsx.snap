// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Verifies that the dialog opens, its content is visible and then matches the snapshot 1`] = `
<div>
  <div
    class="sel-widget"
  >
    <h1>
      Element selection
    </h1>
    <p>
      Number of selected elements: 
      0
    </p>
    <div
      class="sel-widget__selected-opt-cont"
    >
      <p
        class="no-selected-options"
      >
        Selected elements will appear here
      </p>
    </div>
    <br />
    <button
      class="btn btn--green"
    >
      Change my selection
    </button>
    <div
      class="selector-dialog"
    >
      <p>
        Element Selection Dialog
      </p>
      <div
        class="selector-dialog__filters"
      >
        <div>
          <label>
            Search
          </label>
          <input
            data-testid="mainSearchElement"
            placeholder="Start typing..."
            type="search"
          />
        </div>
        <div>
          <label>
            Filter
          </label>
          <select
            data-testid="mainFilterElement"
          >
            <option
              selected=""
              value="0"
            >
              No filters
            </option>
            <option
              value="1"
            >
              Number &gt; 10
            </option>
            <option
              value="2"
            >
              Number &gt; 50
            </option>
            <option
              value="3"
            >
              Number &gt; 100
            </option>
          </select>
        </div>
      </div>
      <div
        class="selector-dialog__entries"
      />
      <p>
        Currently selected elements:
      </p>
      <div
        class="sel-widget__selected-opt-cont"
      >
        <p
          class="no-selected-options"
        >
          Selected elements will appear here
        </p>
      </div>
      <div
        class="selector-dialog__btns"
      >
        <button
          class="btn btn--green"
        >
          Save
        </button>
        <button
          class="btn btn--red"
        >
          Cancel
        </button>
      </div>
    </div>
  </div>
</div>
`;
